File: sip-core/src/main/kotlin/com/basicfu/sip/core/BeanRegistrar.java
package com.basicfu.sip.core;

import com.basicfu.sip.core.annotation.EnableRedis;
import com.basicfu.sip.core.util.RedisUtil;
import org.springframework.beans.factory.support.BeanDefinitionRegistry;
import org.springframework.beans.factory.support.GenericBeanDefinition;
import org.springframework.context.annotation.ImportBeanDefinitionRegistrar;
import org.springframework.core.type.AnnotationMetadata;

import java.util.Map;

public class BeanRegistrar implements ImportBeanDefinitionRegistrar {

    @Override
    public void registerBeanDefinitions(AnnotationMetadata annotationMetadata, BeanDefinitionRegistry beanDefinitionRegistry) {
        Map defaultAttrs = annotationMetadata.getAnnotationAttributes(EnableRedis.class.getName(), true);
        if (defaultAttrs != null) {
            GenericBeanDefinition beanDefinition = new GenericBeanDefinition();
            beanDefinition.setBeanClass(RedisUtil.class);
            beanDefinition.setSynthetic(true);
            beanDefinitionRegistry.registerBeanDefinition("redisUtil", beanDefinition);
        }
    }
}

File: sip-client/src/main/kotlin/com/basicfu/sip/client/common/BeanRegistrar.kt
package com.basicfu.sip.client.common

import com.basicfu.sip.client.controller.SipController
import com.basicfu.sip.client.util.ApiUtil
import org.springframework.beans.factory.support.BeanDefinitionBuilder
import org.springframework.beans.factory.support.BeanDefinitionRegistry
import org.springframework.context.annotation.ImportBeanDefinitionRegistrar
import org.springframework.core.type.AnnotationMetadata

/**
 * @author basicfu
 * @date 2018/7/16
 */
@Suppress("UNCHECKED_CAST")
class BeanRegistrar : ImportBeanDefinitionRegistrar {
    override fun registerBeanDefinitions(metadata: AnnotationMetadata, registry: BeanDefinitionRegistry) {
        //register feign config set request heads
        registry.registerBeanDefinition(
            ApiUtil::class.java.simpleName,
            BeanDefinitionBuilder.genericBeanDefinition(ApiUtil::class.java).beanDefinition
        )
        registry.registerBeanDefinition(
            SipController::class.java.simpleName,
            BeanDefinitionBuilder.genericBeanDefinition(SipController::class.java).beanDefinition
        )
    }
}
